android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt

```diff
@@ -94,6 +94,7 @@ class TranslationsActionTest {
         // Translated state
         assertEquals(translatedEngineState, tabState().translationsState.translationEngineState)
         assertTrue(tabState().translationsState.isTranslated)
+        assertFalse(tabState().translationsState.isExpectedTranslate)
```

Comment:
>Is there ever a case where isTranslated and isExpectedTranslate can both be true? 

Yes, that would be the most typical case. The testing looks weird because `TranslateExpectedAction` wasn't fired during the test, but IRL, if the user had visited a page not in their preferred languages, then it would have fired. (Or not fired, if it was in their preferred language.)

`isTranslated = true` and `isExpectedTranslate = true`
User is on a translated page that was not originally in their preferred language.

`isTranslated = true` and `isExpectedTranslate = false`
User is on a translated page that they translated from the browser menu on a page that is in their preferred language.

`isTranslated = false` and `isExpectedTranslate = true`
User is on an untranslated page that is not in their preferred language.

`isTranslated = false` and `isExpectedTranslate = false`
User is on an untranslated page that is in their preferred language.

>I think my current read of the code is that isExpectedTranslate is always false unless TranslateExpectedAction is dispatched. Is that correct?

That's correct, this patch is mostly trying to undo the state for when it makes sense to do so. I'm going to manually test against https://github.com/mozilla-mobile/firefox-android/pull/5600 before landing to make sure this is fully acting as expected too.

Reply:
This isn't intended as an immediate to do in this PR.

Reply:
> I could see value in have some form of this living in the code

I would also like to add some in-tree docs once some things settle down too. I have a feeling I could draw a helpful state diagram to keep the nuance from getting lost and forgotten. (There is a lot of nuance and I have to keep looking at the Figma and running through all of the possibilities.) 

>I would be curious about your thoughts on trying to consolidate these boolean values into meaningful enum which can better help inform what the state things are in.

For sure! I like this idea a lot, I think it would make mutually exclusive states much more clear and prevent errors too. (e.g., `isTranslateProcessing` and `isRestoreProcessing` should never both be true.)

>It isn't immediately clear how to include isExpectedTranslate so I will acknowledge that this might not yield the results I would be expecting. Overall, I think this would be interesting in terms of providing more clarity about the resulting state at any point due to the variance in the boolean values.

I think mutually exclusive states that won't cause data loss get consolidated (exactly like the examples you provided). The ones that aren't mutually exclusive are ones like `isOfferTranslate` and `isExpectedTranslate`.

To accommodate the ones that are exclusive, maybe we could have a `TranslationsState` extension that then maps to a bigger overall state that takes in exclusive states? I think if I wrote a state diagram, it might be possible to define the state further. Extension and bigger holistic state could be something to think about, but the exclusive translation states would be clear cut. 

I'll file a code health bug to revisit this on!

Reply:
Filed as : [1881804](https://bugzilla.mozilla.org/show_bug.cgi?id=1881804)